backend/utils/gin-context.go

```diff
@@ -30,3 +31,41 @@ func GinContextToContextMiddleware() gin.HandlerFunc {
 		c.Next()
 	}
 }
+
+// GetOrSetContextWithLock gets or sets a value in the context with a lock
+func GetOrSetContextWithLock[T any](ctx context.Context, key string, factory func() (*T, error)) (*T, error) {
+	ginCtx, err := GinCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+	v, ok := ginCtx.Get(key)
+	if ok {
+		r, ok := v.(*T)
+		if !ok {
+			return nil, merry.Errorf("could not cast value to type %T", r)
+		}
+		return r, nil
+	}
+	lock, ok := ginCtx.Get("context-lock-" + key)
+	if !ok {
+		lock = &sync.Mutex{}
```

Comment:
this is not thread safe because two threads can read "context-lock-${key}" at the same time and each create their own lock.

Reply:
In general feel using the gin context here is a bit awkward. Can we do it differently? Maybe add some "locks" struct to manage the locks at request creation. (i.e. like the `keyLocks` example from chatgpt below but on the context instead of global). Or just a global map with more unique keys :)


```golang
import (
	"context"
	"sync"
)

var keyLocks = struct {
	sync.Mutex
	m map[string]*sync.Mutex
}{m: make(map[string]*sync.Mutex)}

// GetOrSetContextWithLock gets or sets a value in the context with a lock
func GetOrSetContextWithLock[T any](ctx context.Context, key string, factory func() (*T, error)) (*T, error) {
	ginCtx, err := GinCtx(ctx)
	if err != nil {
		return nil, err
	}

	// Get the specific key lock in a thread-safe manner
	keyLocks.Lock()
	lock, exists := keyLocks.m[key]
	if !exists {
		lock = &sync.Mutex{}
		keyLocks.m[key] = lock
	}
	keyLocks.Unlock()

	lock.Lock()
	defer lock.Unlock()

	v, ok := ginCtx.Get(key)
	if ok {
		r, ok := v.(*T)
		if !ok {
			return nil, merry.Errorf("could not cast value to type %T", r)
		}
		return r, nil
	}

	r, err := factory()
	if err != nil {
		return nil, err
	}
	ginCtx.Set(key, r)
	return r, nil
}
```

> Here's the breakdown of changes:
- Introduced a global keyLocks variable, a struct that contains a map to store a lock for each key and a mutex to guard the access to this map.
- In the GetOrSetContextWithLock function, use this keyLocks to retrieve or create the lock for the given key in a thread-safe manner. This ensures that you have a unique lock for each key and it's created and accessed in a concurrent-safe way.
- Removed the logic that stored the lock in the ginCtx because we're now managing locks with our keyLocks.